nats.go

```diff
@@ -260,9 +267,27 @@ type Options struct {
 	// to a server that we were already connected to previously.
 	ReconnectWait time.Duration
 
-	// ReconnectJitter sets the upper bound for a random delay added to
-	// ReconnectWait during a reconnect.
-	ReconnectJitter time.Duration
+	// ReconnectJitterNonTLS sets the upper bound for a random delay added to
+	// ReconnectWait during a reconnect when no TLS is used.
+	// Note that any jitter is capped with ReconnectJitterMax.
+	ReconnectJitterNonTLS time.Duration
+
+	// ReconnectJitterTLS sets the upper bound for a random delay added to
+	// ReconnectWait during a reconnect when TLS is used.
+	// Note that any jitter is capped with ReconnectJitterMax.
+	ReconnectJitterTLS time.Duration
+
+	// ReconnectJitterMax caps the maximum jitter.
+	ReconnectJitterMax time.Duration
+
+	// ReconnectJitterCB sets the handler that is invoked after the library went
+	// through the whole list of URLs and failed to reconnect. The number of
+	// attempts and current jitter is passed as arguments to this handler.
+	// It can return a jitter value based on those metrics.
+	// The library will take a random value between 0 and the returned value
+	// and add that to the ReconnectWait duration.
+	// Note that any jitter is capped with ReconnectJitterMax.
```

Comment:
so with this design library users still cant use their own backoff logic, we can only influence the jitter which not what I was suggesting earlier.  It's not nothing and will help, just pointing out that's not what I was suggesting at all.

Reply:
So you want to control both then yes? The most important is to spread out jitter. What the min wait is stops the attempts from happening quickly but changing that and not jitter (unless you already have randomness in how folks got disconnected) will not help smooth reconnect. So for the grpc example above what are the achieving besides the client just not trying as often?

Reply:
It's a true backoff:

 * not as often
 * over time less often
 * capped at a max duration
 * with jitter always

Reply:
Point I was trying to make was that jitter is more important then the delay. I understand we could have both.

Reply:
In example you provided above can you influence jitter?

Reply:
They have a config that lets you influence all the things, max, default, jitter etc

Reply:
ok gotcha. We should do the same although for us tls and jitter should have more of an influence as we discussed.

Reply:
I think your proposed jitter only approach is fine for most and a good default.

I just dont want to make it impossible to get actual full featured backoff if desired, but dont think you need to go and implement one tbh (I'd still want to use my own that my own system configures consistently for all the places I backoff)

Reply:
Understood. And I think we all agree sane defaults is the high level goal but I also want your use case handled well. Maybe we have the following.

`SetJitter` or `SetJitterDelta`

but we also have a combined one for that callback.
`SetReconnectWaitAndJitter(rw, j)`

Reply:
As long as they are reset when we do reconnect etc.

Reply:
sound like a ever growing list of confusing options, what was wrong with 

```go
CustomReconnectDelay(func(n int) time.Duration {
  return backoff.Default.Duration(n)
})```

that simply trusts that duration for the delay?

Reply:
That is fair, use some library for wait and jitter combined. And just make sane defaults for most scenarios.

In your case is your algorithm TLS aware?

Reply:
no - but your defaults could be of course and would be fine for most.

choria cant be run without tls unless you go and recompile it at which point i dont care :)

Reply:
I vote for that simple function. If set, supersede any ReconnectWait option. So we keep the current jitter options (nonTLS and TLS) for cases where the custom option is not set. Doc for CustomReconnectDelay() will clearly state that the library will sleep for given delay and it is recommended that the function returns a value that include some jitter.

Reply:
I am ready to submit updates. Do you guys prefer that I squash first or want to see just the difference (and I will only squash prior to merge)?

Reply:
Squash for me.